### DISCUSSION TITLE
The once() helper in static methods in a parent class.

### DISCUSSION THREAD

**Discussion by bjhijmans** [laravel/framework#51070] [General]
I've been using the new once() helper in Laravel 11, and I found behavior that surprised me.

```php
abstract class Parent
{
    public static function load()
    {
        return once(fn () => File::get(static::$path));
    }
}

class ChildA extends Parent
{
    protected static string $path = 'path1';
}


class ChildB extends Parent
{
    protected static string $path = 'path2';
}
```

Now when I call `ChildA::load()` and then `ChildB::load()`, I get the result from `ChildA` both times because the once hash only takes into account the class that it is in, not the class that it was called from. It puts `Parent::class` into the hash but not `ChildA::class`.

If they weren't static, I'd call `(new ChildA)->load()` and `(new ChildB)->load()` and get the correct results because they are different instances.

There is nothing about this in the documentation (nor, indeed, anything about using once() in static methods at all). I think it should be possible to get the calling class from the callable by reflection.

This behavior feels like a bug to me. I definitely expected it to work differently, but that's my personal opinion. What do you think?

**Comment by henzeb** [comment]
Does this behavior not exist in spatie's once?

**Reply by bjhijmans** [reply]
I never used the Spatie package, but from the code it looks like it does the same thing the Laravel version does.
